src/signTxPoolRegistration.c

```diff
@@ -0,0 +1,878 @@
+			VALIDATE(p->marginNumerator <= MARGIN_DENOMINATOR_MAX, ERR_INVALID_DATA);
+
+			ASSERT_TYPE(p->marginDenominator, uint64_t);
+			p->marginDenominator = u8be_read(wireHeader->marginDenominator);
+			TRACE_BUFFER((uint8_t *) &p->marginDenominator, 8);
+			VALIDATE(p->marginDenominator != 0, ERR_INVALID_DATA);
+			VALIDATE(p->marginDenominator <= MARGIN_DENOMINATOR_MAX, ERR_INVALID_DATA);
+			VALIDATE(p->marginNumerator <= p->marginDenominator, ERR_INVALID_DATA);
+
+			TRACE_BUFFER(wireHeader->rewardAccount, SIZEOF(wireHeader->rewardAccount));
+			ASSERT(SIZEOF(wireHeader->rewardAccount) == SIZEOF(p->rewardAccount));
+			os_memmove(p->rewardAccount, wireHeader->rewardAccount, SIZEOF(p->rewardAccount));
+			const uint8_t header = getAddressHeader(p->rewardAccount, SIZEOF(p->rewardAccount));
+			VALIDATE(getAddressType(header) == REWARD, ERR_INVALID_DATA);
+			VALIDATE(getNetworkId(header) == commonTxData->networkId, ERR_INVALID_DATA);
+		}
+
+		ASSERT_TYPE(subctx->numOwners, uint16_t);
+		ASSERT_TYPE(subctx->numRelays, uint16_t);
+		subctx->numOwners = (uint16_t) u4be_read(wireHeader->numOwners);
+		subctx->numRelays = (uint16_t) u4be_read(wireHeader->numRelays);
+
+		TRACE(
+		        "num owners, relays: %d %d",
+		        subctx->numOwners, subctx->numRelays
+		);
+		VALIDATE(subctx->numOwners <= POOL_MAX_OWNERS, ERR_INVALID_DATA);
+		VALIDATE(subctx->numRelays <= POOL_MAX_RELAYS, ERR_INVALID_DATA);
+
+		// there should be exactly one owner given by path for which we provide a witness
+		VALIDATE(subctx->numOwners >= 1, ERR_INVALID_DATA);
+	}
+
+	// Note: make sure that everything in subctx is initialized properly
+	txHashBuilder_addPoolRegistrationCertificate(
+	        txHashBuilder,
+	        &subctx->poolParams,
+	        subctx->numOwners, subctx->numRelays
+	);
+
+	security_policy_t policy = policyForSignTxCertificateStakePoolRegistration();
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {subctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_PROMPT_BEFORE_RESPONSE, HANDLE_POOLPARAMS_STEP_DISPLAY_OPERATION);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT,   HANDLE_POOLPARAMS_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+
+	handlePoolParams_ui_runStep();
+}
+
+// ============================== OWNER ==============================
+
+enum {
+	HANDLE_OWNER_STEP_DISPLAY = 6320,
+	HANDLE_OWNER_STEP_RESPOND,
+	HANDLE_OWNER_STEP_INVALID,
+};
+
+static void handleOwner_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	ui_callback_fn_t* this_fn = handleOwner_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_OWNER_STEP_DISPLAY) {
+		ui_displayOwnerScreen(&subctx->owner, subctx->currentOwner, commonTxData->networkId, this_fn);
+	}
+	UI_STEP(HANDLE_OWNER_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		subctx->currentOwner++;
+		if (subctx->currentOwner == subctx->numOwners) {
```

Comment:
isn't this included in advanceState? :thinking:

Reply:
My preference is to avoid validation in advanceState, so there are just asserts making sure that an invalid situation has been already properly detected when handling APDUs (in this case, a missing/superfluous path owner).